src/signer.rs

```diff
@@ -0,0 +1,29 @@
+use crate::{Error, Keystore, Message, Signature};
+
+/// Signature provider based on a keystore
+pub trait Signer {
+    /// Concrete keystore implementation
+    type Keystore: Keystore;
+
+    /// Accessor to the keystore
+    fn keystore(&self) -> &Self::Keystore;
+
+    /// Sign a given message with the secret key identified by `id`
+    fn sign(&self, id: <Self::Keystore as Keystore>::KeyId, message: &Message) -> Signature {
```

Comment:
This should probably be a result type, in case an implementation of this trait uses a remote signing which could have I/O errors.

Reply:
That is something to think about. The signature itself is infallible:

https://github.com/FuelLabs/fuel-crypto/blob/92bbdc6b346fe6fe8f8af6bb6f15ccf6ac0d8b71/src/signature.rs#L158

However, the secret key might be fallible:

https://github.com/FuelLabs/fuel-crypto/blob/92bbdc6b346fe6fe8f8af6bb6f15ccf6ac0d8b71/src/keystore.rs#L12

Since we potentially I/O to fetch it. I was thinking of putting an associated type `Error` so we always assume they might be fallible. In case someone implements an in-memory signer, they could just set `Error` to `()`. Thoughts?

Reply:
Alternative, we can have two methods: `sign` and `sign_fallible`. But I don't really like this alternative since it may increase the API complexity

Reply:
Pushed https://github.com/FuelLabs/fuel-crypto/pull/9/commits/4b7228954a3ca20c9583a1f350b96f7b60249839 that will use `Error` as associated.

Introduced `borrown` after searching for some `Cow` similar implementation for no-std, and not finding.